### PR TITLE
#A03. As Admin, I would like to update existing Hackathon events.

### PR DESCRIPTION
**Pull request by hebs87**
## User Story ID:
- #A03 (As Admin, I would like to update existing Hackathon events)
- Feature

## Description of PR:
All changes made are outlined below:
  - Added format attr to start_date and end_date HackathonForm field DateTimeInput widget to render existing datetime value correctly in frontend form when updating hackathon
  - Added update_hackathon view to enable editing hackathon in frontend form - uses create-event.html but pre-populates form fields with existing instance data:
    - If user is not admin user, they are redirected to the hackathons list page
    - Also adds current datetime to Hackathon instance's 'updated' field on save
    - If hackathon has started, allow start date to be a date in the past
    - If hackathon has not yet started and start_date is amended to a date in the past, error message is displayed
    - If end_date is before start_date, error message is displayed
    - If all checks pass, form is submitted and existing instance data is updated with new values
  - Added new URL for the view with the 'hackathon_id' arg to enable the correct event to be updated
  - Added logic to conditionally render 'Edit Hackathon' or 'Create Hacakathon' in create-event.html based on whether user is creating or updating hackathon
  - Amended hackathon_card.html with the following:
    - Added 'EDIT' button (only for admin users) to each rendered hackathon card, which does the following:
      - If hackathon is in progress or ended, it triggers a confirmation modal warning (modal text changes depending on whether hackathon is in progress or ended):
        - Clicking 'NO' closes the modal
        - Clicking 'YES' redirects user to create-event.html template for editing relevant hackathon
      - If hackathon hasn't started yet, the user is taken to the create-event.html template without triggering the warning modal
      - Field values are pre-populated and user will have the ability to update them

## Tests:
Manually tested all implemented functionality, which is all working as expected.
Validated all code in relevant validators and all passed - HTML validator threw errors for Django template tags, but no other errors.

## Know bugs/errors:
No known bugs or errors after testing.

## Screenshots:
image

image

image

image

## Does this introduce a breaking change
- [X] Yes
- [] No

### CODE REVIEW

File: hackathon/views.py [hunk 1]
```diff
@@ -70,6 +70,55 @@ def create_hackathon(request):
...
+    """ Allow users to edit hackathon event """
+
+    # Redirect user if they are not admin
+    if not request.user.is_superuser:
+        return redirect("hackathon:hackathon-list")
+
+    hackathon = get_object_or_404(Hackathon, pk=hackathon_id)
+
+    if request.method == 'GET':
+
```
Comment (stefdworschak) [inline hackathon/views.py:84]: Please remove the unneeded blank line

File: hackathon/urls.py [hunk 2]
```diff
@@ -1,9 +1,10 @@
 from django.urls import path
 
-from .views import HackathonListView, create_hackathon, delete_hackathon
+from .views import HackathonListView, create_hackathon, update_hackathon, delete_hackathon
```
Comment (stefdworschak) [inline hackathon/urls.py:3]: This line is too long as per PEP8 it should be less than 79 chars. Please make sure that all of the lines/files adhere to this.

File: hackathon/views.py [hunk 3]
```diff
@@ -70,6 +70,55 @@ def create_hackathon(request):
...
+            return redirect("hackathon:update_hackathon", hackathon_id)
+
+        # Submit form and save record
+        if form.is_valid():
+            form.instance.updated = now
+            form.save()
+            messages.success(request, f'Thanks, {hackathon.display_name} has been successfully updated!')
+        return redirect("hackathon:hackathon-list")
+
+    pass
```
Comment (stefdworschak) [inline hackathon/views.py:119]: Can you please remove the `pass` (this is not required)  and the unneeded blank line above?

**Review by stefdworschak — APPROVED** [review]
LGTM

@hebs87 please have a quick look at my comments and let me know when I can merge it.

File: hackathon/views.py [hunk 4]
```diff
@@ -70,6 +70,55 @@ def create_hackathon(request):
...
+
+    # Redirect user if they are not admin
+    if not request.user.is_superuser:
+        return redirect("hackathon:hackathon-list")
+
+    hackathon = get_object_or_404(Hackathon, pk=hackathon_id)
+
+    if request.method == 'GET':
+
+        template = "hackathon/create-event.html"
```
[the remark this answers is not in the document]
Comment (hebs87) [reply hackathon/views.py:85]: This is just a personal preference, but if the preference here is to pass it in directly then that's cool. I've made the change, it'll be visible when I push it to the existing PR 👍

File: hackathon/views.py (hunk 1 above)
Comment (hebs87) [reply hackathon/views.py:84]: Done, this will be visible when I push the change to the existing PR.

File: hackathon/urls.py (hunk 2 above)
Comment (hebs87) [reply hackathon/urls.py:3]: Done, this will be visible once I push the changes to the existing PR.

File: hackathon/views.py (hunk 3 above)
Comment (hebs87) [reply hackathon/views.py:119]: Done. This will be visible once I push the changes to the existing PR.

**Review by stefdworschak — APPROVED** [review]
LGTM

**Pull request merged**